lib/media/streaming_engine.js

```diff
@@ -2160,12 +2177,21 @@ shaka.media.StreamingEngine.prototype.clearBuffer_ =
 
   mediaState.waitingToClearBuffer = false;
   mediaState.waitingToFlushBuffer = false;
+  mediaState.deferredClearBufferOffset = 0;
   mediaState.clearingBuffer = true;
 
   shaka.log.debug(logPrefix, 'clearing buffer');
-  let p = this.playerInterface_.mediaSourceEngine.clear(mediaState.type);
+  let p;
+  if (safeMargin) {
+    let playheadTime = this.playerInterface_.playhead.getTime();
+    let duration = this.playerInterface_.mediaSourceEngine.getDuration();
+    p = this.playerInterface_.mediaSourceEngine.remove(mediaState.type,
+        playheadTime + safeMargin, duration);
+  } else {
+    p = this.playerInterface_.mediaSourceEngine.clear(mediaState.type);
+  }
   p.then(function() {
-    if (!this.destroyed_ && flush) {
+    if (!this.destroyed_ && flush && safeMargin == 0) {
```

Comment:
Since this is dependent on `safeMargin == 0`, you could put this as part of line 2191.

Reply:
What did you have in mind here? Not sure how to move this without duplicating things after.
```
p.then(function() {
    if (!this.destroyed_ && flush && safeMargin == 0) {
      return this.playerInterface_.mediaSourceEngine.flush(mediaState.type);
    }
  }.bind(this)).then(function() {
    if (this.destroyed_) return;
    shaka.log.debug(logPrefix, 'cleared buffer');
    mediaState.lastStream = null;
    mediaState.lastSegmentReference = null;
    mediaState.clearingBuffer = false;
    mediaState.endOfStream = false;
    this.scheduleUpdate_(mediaState, 0);
  }.bind(this));
```

Reply:
```js
if (safeMargin) {
  p = ...
} else {
  p = this.playerInterface_.mediaSourceEngine.clear(mediaState.type).then(() => {
    if (!this.destroyed_ && flush)
      return this.playerInterface_.mediaSourceEngine.flush(mediaState.type);
  });
}
p.then(() => {
  if (this.destroyed_) return;
  shaka.log.debug(logPrefix, 'cleared buffer');
  ...
});
```

Reply:
Oh okay, got it, thanks!